File: sms_back_end/database/finace_management/schemas/school_fee_configuration.py
# schemas/finance/school_fee_configuration.py

from datetime import date, datetime
from uuid import UUID
from typing import Optional

from pydantic import BaseModel

from database.finace_management.enum.paymentModules import (
    FeeStructureType
)


# =========================================
# BASE
# =========================================

class SchoolFeeConfigurationBase(BaseModel):

    school_id: UUID

    year_start: date

    year_end: date

    fee_structure: FeeStructureType

    allow_partial_payments: bool = True

    allow_late_payments: bool = True


# =========================================
# CREATE
# =========================================

class SchoolFeeConfigurationCreate(
    SchoolFeeConfigurationBase
):
    pass


# =========================================
# UPDATE
# =========================================

class SchoolFeeConfigurationUpdate(BaseModel):

    year_start: Optional[date] = None

    year_end: Optional[date] = None

    fee_structure: Optional[
        FeeStructureType
    ] = None

    allow_partial_payments: Optional[
        bool
    ] = None

    allow_late_payments: Optional[
        bool
    ] = None


# =========================================
# RESPONSE
# =========================================

class SchoolFeeConfigurationResponse(
    SchoolFeeConfigurationBase
):

    id: UUID

    created_at: datetime

    updated_at: datetime

    model_config = {
        "from_attributes": True
    }

File: sms-frontend/types/finance/schoolFeeConfiguration.ts
// types/finance/endpoint.ts

export type FeeStructureType =
    | 'quarterly'
    | 'monthly'
    | 'annual'

/* =========================================
   BASE
========================================= */

export interface SchoolFeeConfigurationBase {
    school_id: string

    year_start: string

    year_end: string

    fee_structure: FeeStructureType

    allow_partial_payments: boolean

    allow_late_payments: boolean
}

/* =========================================
   CREATE
========================================= */

export type SchoolFeeConfigurationCreate = SchoolFeeConfigurationBase

/* =========================================
   UPDATE
========================================= */

export interface SchoolFeeConfigurationUpdate {
    year_start?: string

    year_end?: string

    fee_structure?: FeeStructureType

    allow_partial_payments?: boolean

    allow_late_payments?: boolean
}

/* =========================================
   RESPONSE
========================================= */

export interface SchoolFeeConfigurationRead
    extends SchoolFeeConfigurationBase {

    id: string

    created_at: string

    updated_at: string
}